File: Java/src/test/java/tree/AvlTest.java
package tree;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.example.tree.AvlTree;

public class AvlTest {

    @Test
    public void testAvlTree() {
        AvlTree tree = new AvlTree();
        for (int i = 0; i <= 150; i += 10) {
            tree.insert(i);
        }

        Assertions.assertEquals(tree.inOrder().toString(),
                "[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150]");
    }

    @Test
    public void testAvlTreeLeftRotate() {
        AvlTree tree = new AvlTree();
        tree.insert(10);
        tree.insert(20);
        tree.insert(30);
        Assertions.assertEquals(tree.inOrder().toString(), "[10, 20, 30]");
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20, 10, 30 });

    }

    @Test
    public void testAvlTreeRightRotate() {
        AvlTree tree = new AvlTree();
        tree.insert(30);
        tree.insert(20);
        tree.insert(10);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20, 10, 30 });
        Assertions.assertArrayEquals(tree.inOrder().toArray(), new Integer[] { 10, 20, 30 });
    }

    @Test
    public void testAvlTreeLeftRightRotate() {
        AvlTree tree = new AvlTree();
        tree.insert(30);
        tree.insert(10);
        tree.insert(20);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20, 10, 30 });
        Assertions.assertArrayEquals(tree.inOrder().toArray(), new Integer[] { 10, 20, 30 });
    }

    @Test
    public void testAvlTreeRightLeftRotate() {
        AvlTree tree = new AvlTree();
        tree.insert(10);
        tree.insert(30);
        tree.insert(20);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20, 10, 30 });
        Assertions.assertArrayEquals(tree.inOrder().toArray(), new Integer[] { 10, 20, 30 });
    }

    @Test
    public void testAvlTreeDelete() {
        AvlTree tree = new AvlTree();
        for (int i = 10; i <= 30; i += 10) {
            tree.insert(i);
        }
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20, 10, 30 });
        tree.delete(30);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20, 10 });

        tree.delete(10);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 20 });

        tree.delete(20);
        Assertions.assertTrue(tree.isEmpty());
    }

    @Test
    public void testAvlTreeDelete2() {
        AvlTree tree = new AvlTree();
        
        tree.insert(10);
        tree.insert(20);
        tree.insert(30);
        tree.insert(40);
        tree.insert(50);
        tree.insert(60);    
        tree.insert(70);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 40, 20, 60, 10, 30, 50,70 });

        tree.delete(40);

        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 50, 20, 60, 10, 30, 70 });

        tree.delete(60);
        Assertions.assertArrayEquals(tree.levelOrder().toArray(), new Integer[] { 50, 20, 70, 10, 30 });
    }

}

File: dart/test/tree/avl_test.dart
import 'package:test/test.dart';

import '../../src/tree/avlTree/avl.dart';

void main() {
  test("test avl tree", () {
    var tree = new AvlTree();

    expect(tree.isEmpty, true);
  });

  test("avl tree insert left rotate", () {
    var tree = new AvlTree();
    tree.insert(10);
    tree.insert(20);
    tree.insert(30);

    expect(
      tree.toList(),
      [20, 10, 30],
    );

    tree.insert(40);
    tree.insert(50);
    expect(
      tree.toList(),
      [20, 10, 40, 30, 50],
    );
  });

  test("avl tree insert right rotate", () {
    var tree = new AvlTree();
    tree.insert(30);
    tree.insert(20);
    tree.insert(10);
    expect(tree.toList(), [20, 10, 30]);

    tree.insert(9);
    tree.insert(8);

    expect(tree.toList(), [20, 9, 30, 8, 10]);
  });

  test("avl tree insert left right rotate", () {
    var tree = new AvlTree();
    tree.insert(30);
    tree.insert(10);
    tree.insert(20);

    expect(tree.toList(), [20, 10, 30]);
  });

  test("avl tree insert right left rotate", () {
    var tree = new AvlTree();
    tree.insert(10);
    tree.insert(30);
    tree.insert(20);
    expect(tree.toList(), [20, 10, 30]);
  });

  test("avl tree delete", () {
    var tree = new AvlTree();
    tree.insert(10);
    tree.insert(20);
    tree.insert(30);
    tree.insert(40);
    tree.insert(50);
    tree.insert(60);
    expect(tree.toList(), [40, 20, 50, 10, 30, 60]);

    tree.delete(60);
    expect(tree.toList(), [40, 20, 50, 10, 30]);

    tree.delete(50);
    expect(tree.toList(), [20, 10, 40, 30]);
    tree.delete(10);
    expect(tree.toList(), [30, 20, 40]);

    tree.insert(25);
    expect(tree.toList(), [30, 20, 40, 25]);

    tree.delete(40);
    expect(tree.toList(), [25, 20, 30]);

    tree.delete(25);
    expect(tree.toList(), [30, 20]);
  });
}
